src/mongo/db/pipeline/expression.cpp

```diff
@@ -2716,31 +2731,110 @@ Value ExpressionIndexOfArray::evaluate(const Document& root) const {
-            return Value(static_cast<int>(i));
+    return ExpressionIndexOfArray::Arguments(vpOperand[1]->evaluate(root), startIndex, endIndex);
+}
+/**
+ * This class handles the case where IndexOfArray is given an ExpressionConstant
+ * instead of using a vector and searching through it we can use a unordered_map
+ * for O(1) lookup time.
+ */
+class ExpressionIndexOfArray::Optimized : public ExpressionIndexOfArray {
+public:
+    Optimized(const boost::intrusive_ptr<ExpressionContext>& expCtx,
+              const ValueUnorderedMap<vector<int>>& indexMap,
+              const ExpressionVector& operands)
+        : ExpressionIndexOfArray(expCtx), _indexMap(std::move(indexMap)) {
+        vpOperand = operands;
+    }
+
+    virtual Value evaluate(const Document& root) const {
+        auto args = evaluateAndValidateArguments(root, vpOperand, _indexMap.size());
+        auto indexVec = _indexMap.find(args.targetOfSearch);
+        if (indexVec != _indexMap.end()) {
+            // Search through the vector of indecies for first index in our range.
+            for (auto index : indexVec->second) {
+                if (index >= args.startIndex && index < args.endIndex) {
+                    return Value(index);
+                }
+            }
+            // the value we are searching for exists but is not in our range.
+            return Value(-1);
+        } else {
+            return Value(-1);
         }
     }
 
-    return Value(-1);
+private:
+    const ValueUnorderedMap<vector<int>> _indexMap;
+};
+
+intrusive_ptr<Expression> ExpressionIndexOfArray::optimize() {
+    // This will optimize all arguments to this expression.
+    auto optimized = ExpressionNary::optimize();
+    if(optimized.get() != this){
+        return optimized;
+    }
+    // If the input array is an ExpressionConstant we can optimize using a unordered_map instead of an
+    // array.
+    if (auto constantArray =  dynamic_cast<ExpressionConstant*>(vpOperand[0].get())) {
+        const Value valueArray = constantArray->getValue();
+        if (valueArray.nullish()) {
+            return ExpressionConstant::create(getExpressionContext(), Value(BSONNULL));
+        }
+        uassert(50749,
+                str::stream() << "First operand of $indexOfArray must be an array. First "
+                              << "argument is of type: "
+                               << typeName(valueArray.getType()),
+                valueArray.isArray());
+
+        auto arr = valueArray.getArray();
+        // To handle the case of duplicate values the values need to map to a vector of indecies.
+        ValueUnorderedMap<vector<int>> indexMap =
+            getExpressionContext()->getValueComparator().makeUnorderedValueMap<vector<int>>();
+
+        for (int i = 0; i < int(arr.size()); i++) {
+            if(indexMap.find(arr[i]) == indexMap.end()){
+               indexMap.emplace(arr[i], vector<int>());
+               indexMap[arr[i]].push_back(i); 
```

Comment:
Since you're doing this in both the if and the else, I'd recommend re-writing lines 2826-2832 as
```
if (indexMap.find(arr[i]) == indexMap.end()) {
    indexMap.emplace(arr[i], vector<int>());
}
indexMap[arr[i]].push_back(i);
```